### PR TITLE
Add template update mechanism via `update_yml()`

### PR DESCRIPTION
**Pull request by pat-s**
- Adds update mechanism via `update_yml()` for GitHub Actions. Once set up, one can simply call `update_yml()` to update to the latest upstream version in {tic} with user changes preserved.
- Add vignette "updating"
- Reorganizes vignette menu structure
- Update precommit hooks (including WORDLIST)
- `use_*_yml()` functions can now optionally return the respective template rather than always writing to disk
- All templates now have a revision date and a "type" tag

## New soft dependencies

- stringr
- ~~lubridate~~

### CODE REVIEW

File: inst/templates/ghactions-meta-macos-windows.yml [hunk 1]
```diff
@@ -20,6 +22,6 @@ jobs:
         config:
           # comment out lines if you do not want to build on certain platforms
           - { os: macOS-latest, r: "release" }
-          - { os: macOS-latest, r: "devel", pkgdown: "true"}
+          - { os: macOS-latest, r: "devel"}
```
Comment (krlmlr) [inline inst/templates/ghactions-meta-macos-windows.yml:25]: Saw this too, skipping further instances.

File: vignettes/updating.Rmd [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+{tic} comes with the convenience of providing YAML templates for various CI providers which remove the need of diving deep into the internals of certain CI providers.
+While these templates work for most packages that use plain R code and do not rely on system libraries (such as GDAL, xml2) or link to other languages (such as Java), R packages which do so often need manual additions in their YAML file.
```
Comment (krlmlr) [inline vignettes/updating.Rmd:18]:
```suggestion
These templates work for most packages that use plain R code.
Packages that rely on system libraries (such as GDAL, xml2) or link to other languages (such as Java) often need manual additions in their YAML file.
```

File: R/update-yaml.R [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+      )
+    } else if (file.exists(template_in)) {
+      template_in <- template_in
+    } else {
+      stop("No template found at supplied path.")
+    }
+
+    # we cannot use {yaml} because the underlying libyaml library does not parse
+    # comments
+    tmpl_local <- readLines(template_in)
```
[the remark this answers is not in the document]
Comment (pat-s) [reply R/update-yaml.R:57]: Could be - haven't tested it. Let's see if problems occur. Thanks for the pointer.

File: R/update-yaml.R [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    tmpl_local[2]
+  ), quiet = TRUE)
+  if (is.na(rev_date_local)) {
+    cli::cli_alert_danger("It looks like your current {.pkg tic} template does
+     not yet have a revision date. Please update the template manually one last
+     time or add a revision date and the template type manually as the  first
+     line of your template.", wrap = TRUE)
+    stopc("No revision date found in current template.")
+  }
+  rev_date_latest <- lubridate::dmy("12/04/2020")
```
[the remark this answers is not in the document]
Comment (pat-s) [reply R/update-yaml.R:89]: Oh yeah this shouldn't be in there - and was there only for testing. Good catch!

File: R/update-yaml.R [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+  if (!rev_date_latest > rev_date_local) {
+    rlang::abort(sprintf(
+      "You already have the latest version of the tmp (%s).",
```
[the remark this answers is not in the document]
Comment (pat-s) [reply R/update-yaml.R:93]: Regression, thanks.

**Comment by pat-s** [comment]
> What is this change to tests/testthat/helper-yaml.R ?

Leftover. Removed.

> I don't fully understand the code that does the patching. Initially I thought this is about marking, with comments, which parts of a template the user is allowed to change:

It is about marking+updating but only for GHA. I am bit confused why you pasted an example from a Travis yml? Did you read the initial comment and the instructions for updating?

Circle CI might work in the same way as GHA - for Travis I need to think a bit more because we cannot have arbitrary stages and redundant keys. But I don't want to hold back everything until there is support for all providers.

Regarding the "pkgdown" env vars: These are intended and not a new change of this PR. Whether or not a pkgdown site is built (and from which runner) is determined by this env var rather than in tic.R. This is also already built into the automatic setup for templating.
The `do_pkgdown()` macro is conditioned on this env var. General conditioning on a runner is not really possible in tic.R because the availability of runners differs, depending on which template was chosen by the user.

File: R/update-yaml.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply R/update-yaml.R:86]: Changed as suggested.

**Comment by pat-s** [comment]
Btw, the whole approach plays nicely with https://twitter.com/ma_salmon/status/1258371754088640512.

This would lead to a full CI-based automation of the {tic} templates.

**Comment by pat-s** [comment]
I thought about it again and will directly add support for all providers to have this off the table.

**Comment by pat-s** [comment]
> I don't fully understand the code that does the patching. Initially I thought this is about marking, with comments, which parts of a template the user is allowed to change:

The main problem are edge cases when users need to make changes in specific stages, e.g `before_install`. 
Key are not allowed to be duplicated and cannot be renamed in Travis. Travis is really restrictive here and feels somewhat limited compared to the other providers.

A custom "apt" block could be easily marked - but the challenge is to account for one-line changes and be aware of iterate row ids changes once one change was appended to the latest template.

**Comment by krlmlr** [comment]
Yeah, it's difficult. I remember stages are replaced completely if they occur in the build matrix, the user will need to do the patching if we amend the definition of a stage.

**Comment by pat-s** [comment]
Tested several days in different repos. Let's do this :)

**Pull request merged**